xarray/core/variable.py

```diff
@@ -94,7 +94,8 @@ def as_variable(obj, name=None):
                             '{}'.format(obj))
     elif utils.is_scalar(obj):
         obj = Variable([], obj)
-    elif getattr(obj, 'name', None) is not None:
+    elif (getattr(obj, 'name', None) is not None and not
```

Comment:
Instead, let's call out the types from which we *can* safely get the name for the dimension (there are likely to be other types that reuse `name`). I think that's `IndexVariable` and `pandas.Index`.